library/network/src/modules/SuSEFirewall.rb

```diff
@@ -403,6 +405,255 @@ def _sf2_to_firewalld_service(service)
       end
     end
 
+    def Read
+      # Do not read it again and again
+      # to avoid rewriting changes already made
+      if @configuration_has_been_read
+        Builtins.y2milestone(
+          "FirewallD configuration has been read already."
+        )
+        return true
+      end
+
+      ReadCurrentConfiguration()
+
+      Builtins.y2milestone(
+        "Firewall configuration has been read: %1.",
+        @SETTINGS
+      )
+      # to read configuration only once
+      @configuration_has_been_read = true
+
+      # Always call NI::Read, bnc #396646
+      NetworkInterfaces.Read
+
+    end
+
+    def ReadCurrentConfiguration
+
+      # We need to start the service before we query the firewalld rules
+      unless IsStarted()
+        StartServices()
+      end
+      # Get all the information from zones and load them to @SETTINGS["zones"]
+      # The following may seem somewhat complicated but it is more efficient to
+      # do only invoke a single firewall-cmd command instead of iterating over
+      # the zones and then using all the different firewall-cmd commands to get
+      # service, port, masquerade etc information from them.
+      all_zone_info = @fwd_api.list_all_zones()
+      # Drop empty lines
+      all_zone_info.reject! { |x| x.empty? }
+      # And now build the hash
+      zone = nil
+      all_zone_info.each do |e|
+        # is it a zone?
+        z = e.split("\s")[0]
+        if GetKnownFirewallZones().include?(z)
+          zone = z
+          next
+        end
+        if @@zone_attributes.any? { |w| e.include?(w.to_s) }
+          attrs = e.split(":\s")
+          attr = attrs[0].lstrip.to_sym
+          # do not bother if empty
+          next if attrs[1].nil?
+          vals = attrs[1].split("\s")
+          # Fix up for masquerade
+          if attr == :masquerade
+            @SETTINGS[zone][attr] = (vals == "no" ? false : true)
+          else
+            vals.each { |x| @SETTINGS[zone][attr] << x }
+          end
+        end
+      end
+
+      @SETTINGS["enable_firewall"] = IsEnabled()
+      @SETTINGS["start_firewall"] = IsStarted()
+      @SETTINGS[:logging] = @fwd_api.get_log_denied()
+
+      true
+    end
+
+    def WriteConfiguration
```

Comment:
it is single thing, but you mix different level of abstraction here. I think good indicators for abstracting some low level stuff into separate private methods is comments block. For example how it can look like see e.g. https://github.com/yast/yast-bootloader/blob/config_files_proto/src/lib/bootloader/grub2base.rb#L91 where more tricky and low level stuff have own method with self explanatory name.